server-timing/load.php

```diff
@@ -0,0 +1,107 @@
+ * In case that no metrics are registered, this is still called on {@see 'wp_loaded'}, so that even then it still fires
+ * its action hooks as expected.
+ *
+ * @since n.e.x.t
+ */
+function perflab_server_timing() {
+	static $server_timing;
+
+	if ( null === $server_timing ) {
+		$server_timing = new Perflab_Server_Timing();
+		add_filter( 'template_include', array( $server_timing, 'on_template_include' ), PHP_INT_MAX );
+	}
+
+	return $server_timing;
+}
+add_action( 'wp_loaded', 'perflab_server_timing' );
+
+/**
+ * Registers a metric to calculate for the Server-Timing header.
+ *
+ * This method must be called before the {@see 'perflab_server_timing_send_header'} hook.
+ *
+ * @since n.e.x.t
+ *
+ * @param string $metric_slug The metric slug.
+ * @param array  $args        {
+ *     Arguments for the metric.
+ *
+ *     @type callable $measure_callback The callback that initiates calculating the metric value. It will receive
+ *                                      the Perflab_Server_Timing_Metric instance as a parameter, in order to set
+ *                                      the value when it has been calculated. Metric values must be provided in
+ *                                      milliseconds.
+ *     @type string   $access_cap       Capability required to view the metric. If this is a public metric, this
+ *                                      needs to be set to "exist".
+ * }
+ */
+function perflab_server_timing_register_metric( $metric_slug, array $args ) {
+	perflab_server_timing()->register_metric( $metric_slug, $args );
+}
+
+/**
+ * Returns whether an output buffer should be used to gather Server-Timing metrics during template rendering.
+ *
+ * @since n.e.x.t
+ *
+ * @return bool True if an output buffer should be used, false otherwise.
+ */
+function perflab_server_timing_use_output_buffer() {
+	return perflab_server_timing()->use_output_buffer();
+}
+
+/**
+ * Wraps a callback (e.g. for an action or filter) to be measured and included in the Server-Timing header.
+ *
+ * @since n.e.x.t
+ *
+ * @param callable $callback    The callback to wrap.
+ * @param string   $metric_slug The metric slug to use within the Server-Timing header.
+ * @param string   $access_cap  Capability required to view the metric. If this is a public metric, this needs to be
+ *                              set to "exist".
+ * @return callable Callback function that will run $callback and measure its execution time once called.
+ */
+function perflab_wrap_server_timing( $callback, $metric_slug, $access_cap ) {
+	return function( ...$callback_args ) use ( $callback, $metric_slug, $access_cap ) {
+		// Gain access to Perflab_Server_Timing_Metric instance.
+		$server_timing_metric = null;
+
+		perflab_server_timing_register_metric(
+			$metric_slug,
+			array(
+				'measure_callback' => function( $metric ) use ( &$server_timing_metric ) {
+					$server_timing_metric = $metric;
+				},
+				'access_cap'       => $access_cap,
+			)
+		);
```

Comment:
@eugene-manuilov Great catch, I didn't consider that. For some context, I wanted to put the registration in the function so that the metric is only registered when the function is actually executed, which I'd like to keep.

But we can definitely find a way to still ensure the metric is only registered once. However, the bigger question now is: What is the expected behavior for its measurement when the function is called more than once?
1. Should only the first call be measured?
2. Should each call be measured and the sum of the duration be reported as the metric value?

WDYT?

Reply:
Yes, these are good questions... Probably it depends on circumstances in which this function is used. I think lets leave this as is for now and if someone faces any problems with it, we will get back to it and think about a better approach.